source/puppet/4.8/env_isolation.md

```diff
@@ -0,0 +1,94 @@
+
+* Modify your existing r10k hook to run the `generate types` command after code deployment.
+* Create and use a script that first runs r10k for an environment, and then runs `generate types` as a postrun command.
+
+If you have enabled environment level purging in r10k, whitelist the `resource_types` folder so that r10k doesn't purge it.
+
+### Results of `generate types`
+
+When you run the `generate types` command, it scans the entire environment for resource type implementations, excluding the resource types shipped in core Puppet. For each resource type implementation it finds, the command generates a corresponding metadata file named after the resource type in the directory `<env-root>/.resource_types`.
+
+It also syncs the files in the `.resource_types` directory so that:
+
+* Types that have been removed in modules are removed from `resource_types`.
+* Types that have been added are added to `resource_types`.
+* Types that have not changed (based on timestamp) are kept as is.
+* Types that have changed (based on timestamp) are overwritten with freshly generated metadata.
+
+The generated metadata files, which have a `.pp` extension, appear in the code directory. (If you are using Puppet Enterprise with Code Manager and file sync, these files appear in both the staging and live code directories.) These generated files are **read-only**: do not delete them, modify them, nor use expressions from them in regular manifests.
+
+## Troubleshooting environment isolation
+
+If `generate types` cannot generate certain types, if the type generated has missing or inaccurate information, or if the generation itself errors or fails, this can cause a catalog compilation error of "type not found" or "attribute not found". These issues can also be caused by manual changes in metadata files, incorrect permissions, or incorrectly implemented resource types.
+
+To diagnose or repair issues:
+
+1. Ensure that your Puppet resource types are correctly implemented. Refactor any problematic resource types.
+
+1. Regenerate the metadata by removing the `<env>/.resource_types` directory and running `generate types` for each environment.
+
+1. If you continue to get catalog compilation errors, disable environment isolation to help you isolate the error.
+
+   * To disable environment isolation in open source Puppet:
+
+     1. Remove the `generate types` command from any r10k hooks.
+     1. Remove the `.resource_types` directory.
+
+  * To disable environment isolation in Puppet Enterprise:
+
+     1. In `/etc/puppetlabs/puppetserver/conf.d/pe-puppet-server.conf`, remove the `pre-commit-hook-commands` setting.
+     1. In [Hiera](./config_intro.html#configure-settings-with-hiera), set `puppet_enterprise::master::puppetserver::pre_commit_hook_commands: []`. 
+     1. On the command line, run `service pe-puppetserver reload`.
+     1. Delete the `.resource_types` directories from your staging code directory (`/etc/puppetlabs/code-staging`).
```

Comment:
Should also deploy the environment(s) after doing this so that the "un-isolated" code goes into effect (is made live).